src/video/types.rs

```diff
@@ -107,15 +109,64 @@ impl VideoSourceType {
 }
 
 impl VideoEncodeType {
-    //TODO: use trait fromstr, check others places
     pub fn from_str(fourcc: &str) -> VideoEncodeType {
         let fourcc = fourcc.to_uppercase();
-        match fourcc.as_str() {
-            "H264" => VideoEncodeType::H264,
-            "MJPG" => VideoEncodeType::Mjpg,
-            "YUYV" => VideoEncodeType::Yuyv,
-            _ => VideoEncodeType::Unknown(fourcc),
+
+        let fourcc = match fourcc.as_str() {
+            "H264" => return VideoEncodeType::H264,
+            "MJPG" => return VideoEncodeType::Mjpg,
+            // TODO: We can implement a [GstDeviceMonitor](https://gstreamer.freedesktop.org/documentation/gstreamer/gstdevicemonitor.html?gi-language=c) and then this manual mapping between v4l's and gst's fourcc will not be neccessary anymore.
+            // A list of possible v4l fourcc from the Linux docs:
+            // - [YUV](https://www.kernel.org/doc/html/v4.8/media/uapi/v4l/yuv-formats.html)
+            // - [RGB](https://www.kernel.org/doc/html/v4.8/media/uapi/v4l/pixfmt-rgb.html)
+            // - [compressed](https://www.kernel.org/doc/html/v4.8/media/uapi/v4l/pixfmt-013.html)
+            "YUYV" => "YUY2".to_string(),
+            _ => fourcc,
+        };
+
+        // Use Gstreamer to recognize raw formats. [GStreamer raw formats](https://gstreamer.freedesktop.org/documentation/additional/design/mediatype-video-raw.html?gi-language=c#formats)
+        use std::ffi::CString;
+        if let Ok(c_char_array) = CString::new(fourcc.clone()).map(|c_str| c_str.into_raw()) {
+            use gst_video::ffi::*;
+
+            let gst_video_format = unsafe { gst_video_format_from_string(c_char_array) };
+
+            if !matches!(
+                gst_video_format,
+                GST_VIDEO_FORMAT_UNKNOWN | GST_VIDEO_FORMAT_ENCODED
+            ) {
+                return VideoEncodeType::Raw(fourcc);
+            }
+        }
+
+        VideoEncodeType::Unknown(fourcc)
+    }
+}
+impl<'de> Deserialize<'de> for VideoEncodeType {
```

Comment:
Can you provide more information about why this implementation is necessary ? looking at the code this is not clear why.

Reply:
> Because of the tricky nature of having two Enum fields both encapsulating Strings, plus to create a validation of what is accepted from GStreamer/V4l, I had to write a custom deserializer.

Like, if we define the `Unknown(String)` as `untagged`, everything different from the other options will be Unknown. If we then add another `untagged` `Raw(String)`, then it goes to who is defined first, and we actually want to use custom logic to define if it is Raw or Unknown.

Reply:
So, it would be nice to have such explanation as a comment.